hyperspy/model.py

```diff
@@ -968,6 +976,225 @@ def _model_function(self, param):
+        k = self._component_data.shape[-2]  # the number of components
+
+        fit = self._component_data.T * self.coefficient_array
+        res = target_signal - fit.sum(-1)  # The residual
+        res_dot = np.dot(res.T, res)
+
+        fit_dot = np.matmul(fit.swapaxes(-2, -1), fit)
+        inv_fit_dot = np.linalg.inv(fit_dot)
+        covariance = (1 / (n - k)) * np.dot(res_dot, inv_fit_dot)
+        return covariance
+
+    def _linear_fitting(self, algorithm="ridge_regression", **kwargs):
+        """
+        Multivariate linear fitting
+
+        Parameters:
+        algorithm: 
+            'ridge_regression' - Default using sklearn
+            'matrix_inversion' - Fallback, fragile
+        """
+        if not self._precomputed_components:
+            not_linear_error = (
+                "Not all free parameters are linear. "
+                "Fit with a "
+                "different optimizer or set non-linear "
+                "`parameters.free = False`. These "
+                "parameters are nonlinear:"
+            )
+            nonlinear_parameters = self.nonlinear_parameters
+            if nonlinear_parameters:
+                raise AttributeError(
+                    not_linear_error
+                    + "\n\t"
+                    + str("\n\t".join(str(para) for para in nonlinear_parameters))
+                )
+            if not self.linear_parameters:
+                raise AttributeError("There are no linear components in this model")
+
+            self._set_linear_parameters_to_one()
+            self._set_p0()
+            n_free_para = len(self.free_parameters)
+            assert n_free_para > 0, "Model does not contain any free components!"
+            channels_signal_shape = np.count_nonzero(self.channel_switches)
+
+            self._component_data = np.zeros((n_free_para, channels_signal_shape))
+            self._component_data_fixed = np.zeros(channels_signal_shape)
+
+            self._set_twinned_lists()
+            for component in self:
+                if component.active:
+                    self._append_component(component)
+        
+            if self._is_multifit:
+                self._precomputed_components = True
+
+        target_signal = self.signal()[np.where(self.channel_switches)]
+
+        if self.signal.metadata.Signal.binned is True:
+            target_signal = target_signal / np.prod(
+                tuple((ax.scale for ax in self.signal.axes_manager.signal_axes))
+            )
+
+        target_signal = target_signal - self._component_data_fixed
+
+        sig1Dshape = np.count_nonzero(self.channel_switches)
+
+        # Reshape what may potentially be Signal2D data into a long Signal1D shape
+        target_signal = target_signal.reshape(sig1Dshape)        
+
+        if not import_sklearn.sklearn_installed or algorithm == "matrix_inversion":
+            if algorithm != "matrix_inversion":
+                warnings.warn(
```

Comment:
Maybe use _logger.warning instead?

Reply:
Yes. I notice that warnings is used a lot of places, when getting a logger and using that is better behaviour? See https://github.com/hyperspy/hyperspy/search?q=warnings.warn&unscoped_q=warnings.warn

Reply:
It's often inconsistent. Warnings will only get raised once in a for loop, while the logger warning will appear each time. So it can depend on the desired behaviour